File: app/src/main/java/org/blitzortung/android/alert/handler/AlertSectorHandler.java
package org.blitzortung.android.alert.handler;

import android.location.Location;
import org.blitzortung.android.alert.AlertParameters;
import org.blitzortung.android.alert.object.AlertSector;
import org.blitzortung.android.alert.object.AlertSectorRange;
import org.blitzortung.android.data.beans.Stroke;

public class AlertSectorHandler {

    private final AlertParameters alertParameters;
    
    private final Location strokeLocation;
    
    private Location location;
    
    private long thresholdTime;

    public AlertSectorHandler(AlertParameters alertParameters) {
        this.alertParameters = alertParameters;
        strokeLocation = new Location("");
    }

    protected void setCheckStrokeParameters(Location location, long thresholdTime) {
        this.location = location;
        this.thresholdTime = thresholdTime;
    }

    protected void checkStroke(AlertSector sector, Stroke stroke) {
        if (sector != null) {
            float distance = calculateDistanceTo(stroke);

            for (AlertSectorRange range : sector.getRanges()) {
                if (distance <= range.getRangeMaximum()) {
                    range.addStroke(stroke);

                    if (stroke.getTimestamp() >= thresholdTime) {
                        sector.updateClosestStrokeDistance(distance);
                    }
                    
                    break;
                }
            }
        }
    }

    private float calculateDistanceTo(Stroke stroke) {
        float distanceInMeters = location.distanceTo(stroke.getLocation(strokeLocation));
        return alertParameters.getMeasurementSystem().calculateDistance(distanceInMeters);
    }

    public long getLatestTimestampWithin(float distanceLimit, AlertSector sector) {
        long latestTimestamp = 0;
        
        for (AlertSectorRange range : sector.getRanges()) {
            if (distanceLimit <= range.getRangeMaximum()) {
                latestTimestamp = Math.max(latestTimestamp, range.getLatestStrokeTimestamp());
            }
        }
        
        return latestTimestamp;
    }
}

File: app/src/main/java/org/blitzortung/android/alert/handler/AlertSectorHandler.kt
package org.blitzortung.android.alert.handler


import android.location.Location
import org.blitzortung.android.alert.handler.ProcessingAlertSector
import org.blitzortung.android.data.beans.Strike
import org.blitzortung.android.util.MeasurementSystem

class AlertSectorHandler {

    private val strikeLocation: Location = Location("")

    private var location: Location? = null

    private var thresholdTime: Long = 0

    fun setCheckStrikeParameters(location: Location, thresholdTime: Long) {
        this.location = location
        this.thresholdTime = thresholdTime
    }

    internal fun checkStrike(sector: ProcessingAlertSector, strike: Strike, measurementSystem: MeasurementSystem) {
        location?.let { location ->
            val distance = calculateDistanceTo(location, strike, measurementSystem)

            if (strike.timestamp >= thresholdTime) {
                sector.ranges.find { r -> distance <= r.rangeMaximum }?.let {
                    it.addStrike(strike);
                    sector.updateClosestStrikeDistance(distance)
                }
            }
        }
    }

    private fun calculateDistanceTo(location: Location, strike: Strike, measurementSystem: MeasurementSystem): Float {
        strikeLocation.longitude = strike.longitude.toDouble()
        strikeLocation.latitude = strike.latitude.toDouble()
        val distanceInMeters = location.distanceTo(strikeLocation)
        return measurementSystem.calculateDistance(distanceInMeters)
    }

}